AutoMad/wordGrabber.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Thu Oct  3 11:21:19 2024
+
+@author: dragon
+"""
+
+import os.path
+
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+
+import madDictionary
+import numpy as np
+import pandas as pd
+
+
+# If modifying these scopes, delete the file token.json.
+SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly","https://www.googleapis.com/auth/documents"]
+
+# the dictionary and story list sheet.
+MADLIB_SHEET_ID = "1GKVnudQK-5QdoxBpt2KceHg6ZooyMYwU7fWAVFAyouw"
+WORDS = "Words!A1:z1000"
+STORIES = "Stories!A1:z1000"
+
+def getCreds():
+    creds = None
+    # The file token.json stores the user's access and refresh tokens, and is
+    # created automatically when the authorization flow completes for the first
+    # time.
+    if os.path.exists("token.json"):
+        creds = Credentials.from_authorized_user_file("token.json", SCOPES)
+    # If there are no (valid) credentials available, let the user log in.
+    if not creds or not creds.valid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = InstalledAppFlow.from_client_secrets_file(
+                "credentials.json", SCOPES
+            )
+            creds = flow.run_local_server(port=0)
+        # Save the credentials for the next run
+        with open("token.json", "w") as token:
+            token.write(creds.to_json())
+    return creds
+
+def getStoryByName(name):
+    result = getGoogleSheet(MADLIB_SHEET_ID,STORIES)    
+    values=result['values']
+    df = pd.DataFrame(values)
+    df.columns = df.iloc[0]
+    df = df[1:]
+    for index,row in df.iterrows():
+        if row['storyname'] == name:
+            return row['urlString']
+    return None
+
+def getStories():
+    result = getGoogleSheet(MADLIB_SHEET_ID,STORIES)
+    values=result['values']
+    df = pd.DataFrame(values)
+    df.columns = df.iloc[0]
+    df = df[1:]
+    return df.storyname
+    
+def getStory(docID):
+    creds = getCreds()
+    try:
+        service = build("docs", "v1", credentials=creds)
+        # Retrieve the documents contents from the Docs service.
+        document = service.documents().get(documentId=docID, includeTabsContent=True).execute()
+        plainText=''
+        allTabs = get_all_tabs(document)
+        for tab in allTabs:
+           # Get the DocumentTab from the generic Tab.
+            document_tab = tab.get('documentTab')
+            doc_content = document_tab.get('body').get('content')
+            plainText += read_structural_elements(doc_content)
+        return plainText
+    
+        
+    except HttpError as err:
+        print(err)
```

Comment:
```suggestion
        return "".join(getTabText(tab) for tab in get_all_tabs(document)
    except HttpError as err:
        print(err)
        
def getTabText(tab):
      # Get the DocumentTab from the generic Tab.
      document_tab = tab.get('documentTab')
      doc_content = document_tab.get('body').get('content')
      return read_structural_elements(doc_content)
```

Let me know if I am pushing the functional/list comprehension style at you too hard. I find things much easier to follow if you are not mutating strings and separating out independent concerns into functions, but if that is not helpful at your current stage I can back off.